xcm/xcm-executor/src/traits/weight.rs

```diff
@@ -76,11 +76,27 @@ pub trait WeightTrader: Sized {
 	}
 }
 
-impl WeightTrader for () {
+#[impl_trait_for_tuples::impl_for_tuples(30)]
+impl WeightTrader for Tuple {
 	fn new() -> Self {
-		()
+		for_tuples!( ( #( Tuple::new() ),* ) )
 	}
-	fn buy_weight(&mut self, _: Weight, _: Assets) -> Result<Assets, Error> {
-		Err(Error::Unimplemented)
+
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, Error> {
+		for_tuples!( #(
+			if let Ok(assets) = Tuple.buy_weight(weight, payment.clone()) {
+				return Ok(assets);
+			}
+		)* );
+		Err(Error::TooExpensive)
+	}
+
+	fn refund_weight(&mut self, weight: Weight) -> Option<MultiAsset> {
+		for_tuples!( #(
+			if let Some(asset) = Tuple.refund_weight(weight) {
+				return Some(asset);
+			}
+		)* );
```

Comment:
does this make sense?

We could buy weight with one asset and return weight with another?

Reply:
From my understanding, the sensible `WeightTrader` impl would be that if weights are not bought here, then it won't refund. This is the case for all current traders like `FixedRateOfFungible` and `UsingComponents`.

This case is covered in the added unit tests `weight_trader_tuple_should_work`.

Reply:
yeah that makes sense

Reply:
That seems to be the case if we assume that weight can only be bought by the native token of a chain. I'm trying to understand if this is still valid for assets that are non-native, e.g. USDC on Statemint/Statemine.

Would it be possible to buy weights using non-native tokens? If so, then a `WeightTrader` tuple may have elements that both accept the non-native token, and when `refund_weight` is called, the first `WeightTrader` in the tuple would refund first, which may not be the correct behaviour.

Reply:
> If so, then a `WeightTrader` tuple may have elements that both accept the non-native token, and when `refund_weight` is called, the first `WeightTrader` in the tuple would refund first, which may not be the correct behaviour.

The tuple impl tries buy/refund weights in the same order. If I understand your example correctly, in this case, the first trader buys weight, so it's fine that it refunds.

For instance, we have a tuple with two non-native traders `(UsdcTrader, UsdtTrader)`, if the assets are:
1. Sufficient USDC. `UsdcTrader` do the buying and refund. `UsdtTrader` won't have the chance to try.
2. Sufficient USDT. `UsdtTrader` buys and refunds. `UsdcTrader` tried both first but failed.
3. Insufficient USDC/USDT. Both trader failed.

Reply:
That seems to carry an assumption that each token has a corresponding `WeightTrader` associated with it, e.g. USDC has a `UsdcTrader` abd USDT has a `UsdtTrader`. Is this relationship guaranteed to always be 1:1? If so, then this logic looks to be sufficient enough.

What I'm more concerned about is when you have multiple `WeightTrader`s that references the same token to trade. I'm imagining that some DeFi protocols may be permissive enough to allow for multiple tokens to be used to pay for fees, and each `WeightTrader` in their use case corresponds to a liquidity pool. Let's say both traders reference USDC as a token that they accept for paying for the weights. You'll then have `Lp1Trader` and `Lp2Trader` in a tuple, and the default behaviour here would then be to always buy and refund with `Lp1Trader` first, before `Lp2Trader` gets queried, which may result in an incorrect behaviour.

Though practically speaking, such a use case is complex enough to justify for a custom implementation of the `WeightTrader` trait on a self-defined type, and I wouldn't expect the developers of such a DeFi protocol to simply use a tuple to wrap the 2 `WeightTrader`s. However, if that's the case, then I think it would be good to document what the default behaviour is when multiple `WeightTrader`s are put inside of a tuple.

Reply:
The tuple impl doesn't assume 1:1 between tokens and traders. It's similar to [`SendXcm` tuple](https://github.com/paritytech/polkadot/blob/83d18b4e2858338f9847397b6e227cedceb9a036/xcm/src/v1/traits.rs#L259) that making no assumption of who sends what kind of messages, but simply try one by one until the msg is successfully sent. I think it's reasonable to make only one of the traders buy/refund weights, as the required weights shouldn't be bought/refunded twice, just like an XCM message should be sent twice. And yeah it's helpful to document this behavior.

I agree that for more complex requirements, ppl should really impl a trader themselves instead of expecting this tuple impl to help. Actually we're planning to impl a DEX trader in Karura, like in your example, and it will be used with other backup traders together.

Reply:
@KiChjang Thanks for helping to add documentations. 👍